### PR TITLE
CREATE2 opcode

### PR DESCRIPTION
**Pull request by mersinvald**
Related issue: #355 
Relies on merging of https://github.com/ETCDEVTeam/etcommon-rs/pull/67

### CODE REVIEW

**Comment by mersinvald** [comment]
Okay, tested it thoroughly with https://github.com/ETCDEVTeam/etcommon-rs/pull/69 this time.

File: src/eval/run/system.rs [hunk 1]
```diff
@@ -79,15 +87,31 @@ pub fn create<M: Memory + Default, P: Patch>(state: &mut State<M, P>, after_gas:
...
+            gas_price: state.context.gas_price,
+            gas_limit: l64_after_gas,
+            value,
+            input: init,
+            action: TransactionAction::Create,
+            nonce: state.account_state.nonce(state.context.address).unwrap(),
+        }
+    } else {
+        pop!(state, salt: H256);
+        let init_hash = Keccak256::digest(&init);
```
Comment (r8d8) [inline src/eval/run/system.rs:94]: Better to provide different names, or chain calls if intermediate values unnecessary as variables.

File: src/transaction.rs [hunk 2]
```diff
@@ -457,6 +457,7 @@ impl<M: Memory + Default, P: Patch> VM for TransactionVM<M, P> {
                 ccode_deposit = match transaction.action {
                     TransactionAction::Call(_) => false,
                     TransactionAction::Create => true,
```
Comment (r8d8) [inline src/transaction.rs]: Why not to use `|` operator?

File: src/eval/run/system.rs [hunk 3]
```diff
@@ -70,6 +70,14 @@ macro_rules! try_balance {
...
 
 pub fn create<M: Memory + Default, P: Patch>(state: &mut State<M, P>, after_gas: Gas) -> Option<Control> {
+    do_create(state, after_gas, false)
+}
+
+pub fn create2<M: Memory + Default, P: Patch>(state: &mut State<M, P>, after_gas: Gas) -> Option<Control> {
+    do_create(state, after_gas, true)
+}
+
+fn do_create<M: Memory + Default, P: Patch>(state: &mut State<M, P>, after_gas: Gas, is_create2: bool) -> Option<Control> {
```
Comment (r8d8) [inline src/eval/run/system.rs]: Can't find any other usages of `do_create`, imho better to use 2 separate functions, especially if used once

File: src/eval/run/system.rs (hunk 1 above)
Comment (mersinvald) [reply src/eval/run/system.rs:94]: It's a common practice in rust to shadow intermediate values if they replace each other and not used separately. Makes no sense to me to make a mess with chained calls here:

```
let init_hash = Keccak256::digest(&init);
let init_hash = M256::from(&init_hash[..]);
```
is better then
```
let init_hash = M256::from(&Keccak256::digest(&init)[..]);
```
The second one is too messy imo.

File: src/transaction.rs (hunk 2 above)
Comment (mersinvald) [reply src/transaction.rs]: Sure

File: src/eval/run/system.rs (hunk 3 above)
Comment (mersinvald) [reply src/eval/run/system.rs]: DRY principal. Code is used 2 times, why copy-paste it when we can abstract it out to a helper function? 
Still there is a way to improve that, will abstract-out only a necessary part.

Comment (mersinvald) [reply src/eval/run/system.rs]: Ah, I get what you mean. You're right, we can use the single function and call it with boolean parameter, it's not a public API anyway.

Comment (mersinvald) [reply src/eval/run/system.rs]: Done.

File: src/eval/run/system.rs (hunk 1 above)
Comment (r8d8) [reply src/eval/run/system.rs:94]: ok, lets leave it

**Review by r8d8 — APPROVED** [review]
LGTM

**Pull request merged**